### PR TITLE
test: introduce UI tests with Cypress

### LINKED ISSUE #571: Implement first set of simple UI tests and run in CI 🧪
**Is your feature request related to a problem? Please describe.**

Currently, we do not have any UI tests that run in our CI pipeline. This results in issues like #570, where we are unsure whether a pull request breaks the user interface.

**Describe the solution you'd like**

1. Choose a frontend/UI testing library - Jest appears to be the most common for testing React
2. Write some simple tests to get the ball rolling; i.e. each page of the UI site renders as expected (snapshot testing is quite good for this)
3. Run these tests in the existing CI pipeline (editable in `.github/workflows`)

### PR DESCRIPTION
**Pull request by bill-n**
Closes #571 

Implemented first set of simple UI test using cypress
cypress

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -54,3 +54,19 @@ jobs:
...
     #   run: exit ${{ steps.test.outputs.exit_code }}
+
+    - name: Cypress checkout
+      uses: actions/checkout@v2
+
+    - name: Set up Node.js
+      uses: actions/setup-node@v4
+    
+    - name: Install dependencies
+      run: npm i
```
Comment (JamieSlome) [inline .github/workflows/ci.yml:65]: Do we need these three steps if they are already executed above? 🤔

Comment (bill-n) [reply .github/workflows/ci.yml:65]: It is redundant...I will update the PR

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -53,4 +53,11 @@ jobs:
         token: ${{ secrets.CODECOV_TOKEN }}
     # - name: Exit if coverage condition not met
     #   if: ${{ steps.test.outputs.exit_code }} != 0
-    #   run: exit ${{ steps.test.outputs.exit_code }}
+    #   run: exit ${{ steps.test.outputs.exit_code }
```
Comment (JamieSlome) [inline .github/workflows/ci.yml:56]:
```suggestion
    #   run: exit ${{ steps.test.outputs.exit_code }}
```

**Review by JamieSlome — APPROVED** [review]
LGTM! 🍰

**Comment by JamieSlome** [comment]
@bill-n - hmm, I might have been mistaken (sorry!). It seems like Cypress can't detect the UI anymore? I wonder if we try reverting or add the steps `npm run start` and `npm run build`? Then `localhost:3000` should be exposed for the test suite to run against.

Let me know if you agree 👍

**Comment by JamieSlome** [comment]
@bill-n - do we still require the 'wait-on' dependency?

**Comment by bill-n** [comment]
> @bill-n - do we still require the 'wait-on' dependency?

I included the wait-on dependency because cypress may try to run tests against a server that hasn't fully started, resulting in connection errors and test failures.
In your opinion, do you think we are better off without the wait-on dependency?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -54,3 +54,32 @@ jobs:
     # - name: Exit if coverage condition not met
     #   if: ${{ steps.test.outputs.exit_code }} != 0
     #   run: exit ${{ steps.test.outputs.exit_code }}
+    
+    - name: Cypress install
+      uses: cypress-io/github-action@v6
+      with:
+        runTests: false
+        build: npm run build
```
Comment (JamieSlome) [inline .github/workflows/ci.yml:62]: Shouldn't this step just be `npm run build` by itself? Is the Cypress action required here?

**Comment by JamieSlome** [comment]
@bill-n - are you able to resolve the merge conflict? Keen to get this PR merged as it adds a lot of value to the project and gives us stability reassurance for the UI that we really need 👍

Thank you for your work @bill-n ❤️

**Comment by bill-n** [comment]
> @bill-n - are you able to resolve the merge conflict? Keen to get this PR merged as it adds a lot of value to the project and gives us stability reassurance for the UI that we really need 👍
> 
> Thank you for your work @bill-n ❤️

@JamieSlome merge conflict resolved...thank you

**Comment by JamieSlome** [comment]
@bill-n - thank you for your contribution 👏

**Pull request merged**